crates/ruff_python_formatter/src/context.rs

```diff
@@ -56,7 +67,30 @@ impl Debug for PyFormatContext<'_> {
         f.debug_struct("PyFormatContext")
             .field("options", &self.options)
             .field("comments", &self.comments.debug(self.source_code()))
+            .field("node_level", &self.node_level)
             .field("source", &self.contents)
             .finish()
     }
 }
+
+/// What's the enclosing level of the outer node.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+pub(crate) enum NodeLevel {
```

Comment:
I'm not super happy with this because it will require that any code that formats an indented block needs to set the level to `NodeLevel::Statement`. 

The reason why I introduced it is because the formatting logic needs to know whether two empty lines are acceptable or not (or any, in the case of parenthesized expressions)

Reply:
Will we need even more granular states here? E.g., Black has different rules for docstrings within classes vs. docstrings within functions (requires one newline for the former, _allows_ one newline for the latter):

```py
def f():
    """foo"""
    x = 1


class F:
    """foo"""

    x = 1
```

Reply:
Possibly, but I would prefer adding global state when possible (it comes at a cost that we need to snapshot the state and restore it if we ever add error handling support to the formatter). 

Away we could do docstring formatting without adding any additional global state is to:

* Get the first statement
* Test if it is a docstring
* If so, format it, and insert the appropriate number of newlines after
* Otherwise format it as usual